File: 47. Permutations II.py
class Solution:
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if len(nums) == 0:
            return []
        if len(nums) == 1:
            return [nums]
        ans = [()]
        for i in range(len(nums)):
            le = len(ans)
            l = 0
            while l < le:
                l1 = list(ans.pop(0))
                l1.insert(0, nums[i])
                if tuple(l1) not in ans:
                    ans.append(tuple(l1))
                for j in range(len(l1) - 1):
                    l1[j], l1[j+1] = l1[j+1], l1[j]
                    if tuple(l1) not in ans:
                        ans.append(tuple(l1))
                l += 1
        return [list(x) for x in ans]


sol = Solution()
print(sol.permuteUnique([]))
print(sol.permuteUnique([11]))
print(sol.permuteUnique([11, 11, 22]))
print(sol.permuteUnique([11, 11, 22, 11]))
print(sol.permuteUnique([11, 22, 33]))
print(sol.permuteUnique([11, 22, 33, 22, 11]))

File: 47. Permutations II.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<bool> flag(nums.size(), false);
        vector<vector<int>> res;
        vector<int> curr;
        backtracking(nums, curr, flag, res);
        return res;
    }

    void backtracking(const vector<int> &nums, vector<int> &curr, vector<bool> &flag, vector<vector<int>> &res) {
        if (curr.size() == nums.size()) {
            res.push_back(curr);
            return;
        }
        for (int i=0; i < nums.size(); ++i) {
            if (flag[i] || (i > 0 && nums[i-1] == nums[i] && !flag[i-1])) continue;
            curr.push_back(nums[i]);
            flag[i] = true;
            backtracking(nums, curr, flag, res);
            curr.pop_back();
            flag[i] = false;
        }
    }
};

int main() {
    Solution a;
    vector<int> v{1,2,3,3};
    vector<vector<int>> ans = a.permuteUnique(v);
    for (auto &vec: ans) {
        for (auto element: vec) {
            cout << element << ' ';
        }
        cout << endl;
    }
    return 0;
}
